### PR TITLE
SOLR-12325 Added uniqueBlockQuery(parent:true) aggregation for JSON Facet

### PR DESCRIPTION
**Pull request by a-siuniaev**
# Description

Added uniqueBlockQuery(parent:true) aggregation for JSON Facet, which is an optimized alternative to uniqueBlock(_root_). 

# Solution

Is implemented following the same type of strategy as uniqueBlock, but in this case we check next bit in parents' bitset  - which is faster than getting _root_ values. 

# Tests

Covered by the tests analogous to uniqueBlock tests. 

# Checklist

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/ValueSourceParser.java [hunk 1]
```diff
@@ -975,6 +976,13 @@ public ValueSource parse(FunctionQParser fp) throws SyntaxError {
       }
     });
 
+    addParser("agg_uniqueBlockQuery", new ValueSourceParser() {
+      @Override
+      public ValueSource parse(FunctionQParser fp) throws SyntaxError {
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/search/ValueSourceParser.java]: Hello, @a-siuniaev. Thanks for the patch. Wouldn't you mind to explore [Yonik's idea](https://issues.apache.org/jira/browse/SOLR-12325?focusedCommentId=16649083&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16649083) about  overloading the same name with the different argument type (query)

File: solr/core/src/java/org/apache/solr/search/facet/UniqueBlockQueryAgg.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Objects;
+import java.util.function.IntFunction;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.BitSet;
+
+import static org.apache.solr.search.join.BlockJoinParentQParser.getCachedFilter;
+
+public class UniqueBlockQueryAgg extends AggValueSource {
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/search/facet/UniqueBlockQueryAgg.java]: This reminds me uniqueBlock's code. Can you unify them somehow via extracting ancestors probably?

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java [hunk 3]
```diff
@@ -3183,6 +3183,41 @@ public void testUniquesForMethod() throws Exception {
...
+            "  }" +
+            "}" )
+
+        , "response=={numFound:6,start:0,docs:[]}"
+        , "facets=={ count:6," +
+            "types:{" +
+            "    buckets:[ {val:page, count:6, in_books:2} ]}" +
+            "pages:{" +
+            "    buckets:[ " +
+            "     {val:a, count:1, in_books:1}," +
```
Comment (mkhludnev) [inline solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java]: asserting equal counts across children and parents is little bit pointless. The later, should be smaller than former.

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetsWithNestedObjects.java [hunk 4]
```diff
@@ -360,4 +360,30 @@ public void testUniqueBlock() throws Exception {
...
+            + "  limit:-1," +
+            "    facet: {" +
+            "           in_books: \"uniqueBlockQuery(type_s:book)\" }" +
+            "  }" +
+            "}" )
+
+        , "response=={numFound:2,start:0,docs:[]}"
+        , "facets=={ count:2," +
+            "types:{" +
+            "    buckets:[ {val:review,    count:5, in_books:2} ]}" +
```
Comment (mkhludnev) [inline solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetsWithNestedObjects.java]: This kind of smoke test is fine. I;m a little bit anxious from the code overall. Can you think about bigger scale random test making side by side comparison? Thanks.

File: solr/core/src/java/org/apache/solr/search/ValueSourceParser.java (hunk 1 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/search/ValueSourceParser.java]: Added parseNamedArg to FunctionQParser.

File: solr/core/src/java/org/apache/solr/search/facet/UniqueBlockQueryAgg.java (hunk 2 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/search/facet/UniqueBlockQueryAgg.java]: Extracted.

File: solr/core/src/java/org/apache/solr/search/FunctionQParser.java [hunk 5]
```diff
@@ -231,7 +232,59 @@ public String parseArg() throws SyntaxError {
...
 
-  
+  /**
+   * Parse argument in the same way as parseArg() but also parse the argument name
+   * if written in following syntax: argumentName=argument.
+   *
+   * @return Immutable entry with name as the key and argument as value. In case where there's no name, key is null.
+   * @throws SyntaxError in case when argument is not ended by ) or ,
+   */
+  public SimpleImmutableEntry<String, String> parseNamedArg() throws SyntaxError {
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/search/FunctionQParser.java]: 1. Usually it's nicer to declare interface ie Entry<>.
1. I've thought that Yonik whiteness about something already existed? Doesn't it?

File: solr/core/src/java/org/apache/solr/search/facet/UniqueBlockFieldAgg.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ */
+package org.apache.solr.search.facet;
+
+import java.io.IOException;
+
+import org.apache.solr.schema.SchemaField;
+
+public class UniqueBlockFieldAgg extends UniqueBlockAgg {
+
+  private static final class UniqueBlockFieldSlotAcc extends UniqueBlockSlotAcc {
```
Comment (mkhludnev) [inline solr/core/src/java/org/apache/solr/search/facet/UniqueBlockFieldAgg.java]: is it really necessary?

File: solr/core/src/java/org/apache/solr/search/FunctionQParser.java (hunk 5 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/search/FunctionQParser.java]: 1. Agree. 
2. [Here](https://issues.apache.org/jira/browse/SOLR-12325?focusedCommentId=16649083&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16649083) he said "I wish we had..." - i.e. it did't exist at least at that time. And I can't find it now...

File: solr/core/src/java/org/apache/solr/search/facet/UniqueBlockFieldAgg.java (hunk 6 above)
Comment (a-siuniaev) [reply solr/core/src/java/org/apache/solr/search/facet/UniqueBlockFieldAgg.java]: Agree.

**Pull request closed without merging**